pkg/operator/controllers/pullsecret/pullsecret_controller.go

```diff
@@ -59,87 +73,62 @@ func (r *PullSecretReconciler) Reconcile(request ctrl.Request) (ctrl.Result, err
 		return reconcile.Result{}, nil
 	}
 
-	mysec, err := r.kubernetescli.CoreV1().Secrets(operator.Namespace).Get(ctx, operator.SecretName, metav1.GetOptions{})
+	operatorSecret, err := r.kubernetescli.CoreV1().Secrets(operator.Namespace).Get(ctx, operator.SecretName, metav1.GetOptions{})
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		ps, isCreate, err := r.pullsecret(ctx)
+	var userSecret *corev1.Secret
+
+	// reconcile global pull secret
+	// detects if the global pull secret is broken and fixes it by using backup managed by ARO operator
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		userSecret, err = r.pullsecret(ctx)
 		if err != nil {
 			return err
 		}
-
-		if ps.Data == nil {
-			ps.Data = map[string][]byte{}
-		}
-
-		// validate
-		if !json.Valid(ps.Data[corev1.DockerConfigJsonKey]) {
-			r.log.Info("pull secret is not valid json - recreating")
-			delete(ps.Data, corev1.DockerConfigJsonKey)
+		var action PullSecretAction
+		// fix pull secret if its broken to have at least the ARO pull secret
+		userSecret, action, err = r.updateGlobalPullSecret(operatorSecret, userSecret)
+		if err != nil {
+			return err
 		}
-
-		pullsec, changed, err := pullsecret.Merge(string(ps.Data[corev1.DockerConfigJsonKey]), string(mysec.Data[corev1.DockerConfigJsonKey]))
+		err = r.emitGlobalPullSecretChange(ctx, userSecret, action)
```

Comment:
This is not emitting anything. It does the fixing itself.

Reply:
I get what you doing with the IOTA status but would this not be simple if we just make it single function and handle everything inside? 
It just has to :
1. Get secret -> `isCreate` (pull secret already used to return you a skeleton for the secret)
2. Validate skeleton/secret/merge -> `changed`
3. Update.

Basically just move flow we have in current to separate function.

If we ever get into more state based functions `fixPullSecret`, `emitKeysConditions` we are not able to do this with multiple functions so we will have to refactor again.